### ISSUE TITLE
Cron schedules stop after a certain time

### ISSUE THREAD

**Issue by rlamarche** [redisson/redisson#1112]
Hi,

I'm using cron schedules on an executor service but the schedules stop after a certain amount of time.

I've reproduced the issue in a small standalone project.

Here the scenario ;

- Use latest redisson (3.5.4) with single node redis and using linux native epoll in configuration
- Create a single worker for an executor service
- Create the corresponding executor service
- Schedule a cron task every seconds
- Schedule a cron task every days at midnight

After a certain amount of time, the cron task every seconds will stop working.
When midnight comes, the midnight cron task works, and the every seconds task resume, but then stop again after a certain time.
I did not wait midnight on the standalone project, but I have had this behaviour in my project.

I looked into redis and saw that the every seconds task is still here, and the zset contains valid value for the next run in the score.

Note that I also reproduced the bug with every 5 seconds or every 5 minutes task and the task every days at midnight, but I had to wait longer.

I reproduced also without using the native linux epoll.

Redis version : 3.2.9 from docker official image redis:3.2.9 after 1 to 2 minutes
Redis version : 4.0.2 from docker official image redis:4.0.2 after around 8 minutes

[redisson-issue.log](https://github.com/redisson/redisson/files/1407612/redisson-issue.log)
[redisson-issue.tar.gz](https://github.com/redisson/redisson/files/1407613/redisson-issue.tar.gz)

**Comment by jackygurui** [comment]
Thank for sharing the code. I will try to reproduce it.

**Comment by jackygurui** [comment]
Not been able to reproduce the issue. See my log

https://gist.github.com/jackygurui/e1dd562391742cf1b33382840be6dfaf

**Comment by rlamarche** [comment]
Ok thank you.
Which version of redis are you using ? Do you have a specific configuration ?

When starting redis, I get the following warning :

```
1:M 24 Oct 05:42:22.225 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
1:M 24 Oct 05:42:22.225 # Server initialized
1:M 24 Oct 05:42:22.225 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
1:M 24 Oct 05:42:22.225 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
```

Do you think this can be the cause ?

I reproduce the bug on my machine, and on google's server on a kubernetes cluster.

**Comment by jackygurui** [comment]
the problem you have described doesn't sound like it is related to redis.

Just wondering are you running everything in docker?

**Comment by rlamarche** [comment]
The scenario I've sent you, no (redis in docker, java from IDE on my host), but in my prod environment : yes.

I'm gonna fix the warnings I have in redis and try to reproduce the bug.

**Comment by rlamarche** [comment]
Hi,

I fixed redis warnings and I succeed reproducing the issue.

Maybe it comes from redis, or not, but can you please give a direction to search ?
I looked into the source code the scheduler and I found that it uses "timeouts" with netty.

I've seen 2 possible problems :
- redis is not transmiting a queue event
- netty is not waking up correctly after the given timeout

Here the log : https://gist.github.com/rlamarche/c6b853bd8dc66673f42d877f001717fa

I used a "every 5 minutes" and "every seconds" task. You see in the log that when the "every seconds" task stop working, it resumes with the "every 5 minutes" task.

Thank you.

**Comment by rlamarche** [comment]
Intersting passage in my log :

```
2017-10-26T12:15:00.090+0200 [redisson-6-4] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-26T12:15:00.095+0200 [redisson-6-5] INFO sandbox.SimpleTask - Execute task: ##### Every 5 minutes task
2017-10-26T12:15:01.089+0200 [redisson-6-1] INFO sandbox.SimpleTask - Execute task: Every seconds task
```

**Comment by jackygurui** [comment]
This problem you have reproduced is it in your prod environment (everything in docker / kubernetes)?

**Comment by jackygurui** [comment]
Have you tried it without epoll?

**Comment by rlamarche** [comment]
No, on my computer with redis inside a docker container and java running from IntelliJ Idea.
Yes I did try without epoll but I will try again with this new scenario.

**Comment by jackygurui** [comment]
please try it again with `-server` JVM option

**Comment by rlamarche** [comment]
Same problem with `-server` :-(

**Comment by rlamarche** [comment]
I think that the problem is with the `HashedWheelTimer` from Netty.
Look at this portion of code : (from `QueueTransferTask`) :
I added some logs.

```
        long delay = startTime - System.currentTimeMillis();
        if (delay > 10) {
            log.info("Delay: {}", delay);
            Timeout timeout = connectionManager.newTimeout(new TimerTask() {                    
                @Override
                public void run(Timeout timeout) throws Exception {
                    log.info("Push task with delay: {}", delay);
                    pushTask();
                }
            }, delay, TimeUnit.MILLISECONDS);
            timeoutReference.set(timeout);
        } else {
            pushTask();
        }
```

And the last logs just before the every seconds task stop :


```
2017-10-31T11:11:25.060+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Push task with delay: 939
2017-10-31T11:11:25.061+0100 [redisson-netty-4-2] INFO org.redisson.QueueTransferTask - Delay: 46114939
2017-10-31T11:11:25.061+0100 [redisson-5-5] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T11:11:25.061+0100 [redisson-5-3] INFO org.redisson.QueueTransferTask - Delay: 939
2017-10-31T11:11:26.060+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Push task with delay: 939
2017-10-31T11:11:26.061+0100 [redisson-netty-4-5] INFO org.redisson.QueueTransferTask - Delay: 46113939
2017-10-31T11:11:26.061+0100 [redisson-5-4] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T11:11:26.061+0100 [redisson-5-6] INFO org.redisson.QueueTransferTask - Delay: 939
2017-10-31T11:11:27.060+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Push task with delay: 939
2017-10-31T11:11:27.061+0100 [redisson-5-2] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T11:11:27.062+0100 [redisson-5-8] INFO org.redisson.QueueTransferTask - Delay: 938
2017-10-31T11:11:27.062+0100 [redisson-netty-4-6] INFO org.redisson.QueueTransferTask - Delay: 46112938
```
The log `Delay: 46113939` is from the every days at midnight task. So it is never pushed because I don't wait until midnight.

The log `Delay: 938` is from the every seconds task.

In the same "second", when the short delay is used to add a timeout task **after** the long delay, it works.
But when the short delay is used to add a timeout task occurs **before** the long delay, it don't wake up.
It will wake up **after** the long delay, so after midnight.

Note that the short and the long delay does not occur in the same thread. I don't really know how work the netty scheduler, but I don't think that it relies on redis.

What do you think of it ?

**Comment by rlamarche** [comment]
I don't understand why the long delay create a timeout every seconds. Maybe it interfers with the `HashedWheelTimer`from netty which is "approximated" :

https://netty.io/4.0/api/io/netty/util/HashedWheelTimer.html

**Comment by jackygurui** [comment]
Can you make a test case that could reliably reproduce this issue? So far I can't reproduce it with the code you shared.

**Comment by rlamarche** [comment]
The problem is that netty use platform-dependant code. On which platform are you running my testcase that reproduce the bug every times on my computer (Linux Ubuntu 16.04.3 LTS Linux 4.4.0-97-generic x86_64

In netty HashedWheelTimer code :

```
    private final Queue<HashedWheelTimeout> timeouts = PlatformDependent.newMpscQueue();
    private final Queue<HashedWheelTimeout> cancelledTimeouts = PlatformDependent.newMpscQueue();
```
I will continue to investigate to find where the bug comes from.

**Comment by rlamarche** [comment]
I think I've found the problem.
Look at these logs :

```
2017-10-31T15:45:15.066+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Push task with delay: 930
2017-10-31T15:45:15.067+0100 [redisson-netty-4-3] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 66954277 ns ago, task: org.redisson.QueueTransferTask$3@33166d37)
2017-10-31T15:45:15.067+0100 [redisson-netty-4-3] INFO org.redisson.QueueTransferTask - Delay: 29684933
2017-10-31T15:45:15.068+0100 [redisson-5-3] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T15:45:15.069+0100 [redisson-5-4] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 29684930810155 ns later, task: org.redisson.QueueTransferTask$3@472237dc)
2017-10-31T15:45:15.069+0100 [redisson-5-4] INFO org.redisson.QueueTransferTask - Delay: 931
2017-10-31T15:45:16.065+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Push task with delay: 931
2017-10-31T15:45:16.068+0100 [redisson-5-5] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T15:45:16.069+0100 [redisson-5-6] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 69223015 ns ago, task: org.redisson.QueueTransferTask$3@1b09c38e)
2017-10-31T15:45:16.069+0100 [redisson-5-6] INFO org.redisson.QueueTransferTask - Delay: 931
2017-10-31T15:45:16.069+0100 [redisson-netty-4-6] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 930857019 ns later, task: org.redisson.QueueTransferTask$3@1b999b22)
2017-10-31T15:45:16.070+0100 [redisson-netty-4-6] INFO org.redisson.QueueTransferTask - Delay: 29683931
```

I added a log on the cancel timeout of the method `scheduleTask` in class `QueueTransferTask` :

```
    private void scheduleTask(final Long startTime) {
        if (startTime == null) {
            return;
        }
        
        Timeout oldTimeout = timeoutReference.get();
        if (oldTimeout != null) {
            log.info("Cancel timeout: {}", oldTimeout);
            oldTimeout.cancel();
            timeoutReference.compareAndSet(oldTimeout, null);
        }
        
        long delay = startTime - System.currentTimeMillis();
        if (delay > 10) {
            log.info("Delay: {}", delay);
            Timeout timeout = connectionManager.newTimeout(new TimerTask() {                    
                @Override
                public void run(Timeout timeout) throws Exception {
                    log.info("Push task with delay: {}", delay);
                    pushTask();
                }
            }, delay, TimeUnit.MILLISECONDS);
            timeoutReference.set(timeout);
        } else {
            pushTask();
        }
    }
```

There is only one instance of `QueueTransferTask` for a single worker, and it as a final field `private final AtomicReference<Timeout> timeoutReference = new AtomicReference<Timeout>()` containing the last timeout set.

Because I have two cron tasks on the same executor with 1 worker, the object contained in the `AtomicReference` is the next 1s task, and it is canceled when a timeout for the midnight task is added.

So the problem is that the QueueTransferTask does not receive tasks in ad-hoc order in special conditions. I've not yet explained what are these special conditions but maybe you can help me on this because I don't really know how exactly works redisson :thinking:

**Comment by rlamarche** [comment]
In the next-to-last log, the deadline correspond to the every second task. On all previous task canceled, the deadline correspond to the every days at midnight task.

```
[redisson-netty-4-6] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 930857019 ns later, task: org.redisson.QueueTransferTask$3@1b999b22)
```

**Comment by rlamarche** [comment]
Here some interesting logs:

```
2017-10-31T16:38:08.031+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Timeout expired: HashedWheelTimer$HashedWheelTimeout(deadline: 30669572 ns ago, task: org.redisson.QueueTransferTask$3@43eaff97)
2017-10-31T16:38:08.031+0100 [redisson-netty-4-4] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 31320744 ns ago, task: org.redisson.QueueTransferTask$3@43eaff97)
2017-10-31T16:38:08.031+0100 [redisson-netty-4-4] INFO org.redisson.QueueTransferTask - New timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 26511968998717 ns later, task: org.redisson.QueueTransferTask$3@1f0a4dfd)
2017-10-31T16:38:08.032+0100 [redisson-5-1] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T16:38:08.032+0100 [redisson-5-8] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 26511967865011 ns later, task: org.redisson.QueueTransferTask$3@1f0a4dfd)
2017-10-31T16:38:08.032+0100 [redisson-5-8] INFO org.redisson.QueueTransferTask - New timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 967997802 ns later, task: org.redisson.QueueTransferTask$3@2122bf93)



2017-10-31T16:38:09.031+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Timeout expired: HashedWheelTimer$HashedWheelTimeout(deadline: 30588302 ns ago, task: org.redisson.QueueTransferTask$3@2122bf93)
2017-10-31T16:38:09.032+0100 [redisson-netty-4-5] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 31245161 ns ago, task: org.redisson.QueueTransferTask$3@2122bf93)
2017-10-31T16:38:09.032+0100 [redisson-netty-4-5] INFO org.redisson.QueueTransferTask - New timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 26510967997672 ns later, task: org.redisson.QueueTransferTask$3@289a5a9f)
2017-10-31T16:38:09.032+0100 [redisson-5-4] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T16:38:09.033+0100 [redisson-5-6] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 26510966753649 ns later, task: org.redisson.QueueTransferTask$3@289a5a9f)
2017-10-31T16:38:09.033+0100 [redisson-5-6] INFO org.redisson.QueueTransferTask - New timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 966998071 ns later, task: org.redisson.QueueTransferTask$3@3eb91594)



2017-10-31T16:38:10.031+0100 [pool-2-thread-1] INFO org.redisson.QueueTransferTask - Timeout expired: HashedWheelTimer$HashedWheelTimeout(deadline: 31214896 ns ago, task: org.redisson.QueueTransferTask$3@3eb91594)
2017-10-31T16:38:10.033+0100 [redisson-5-5] INFO sandbox.SimpleTask - Execute task: Every seconds task
2017-10-31T16:38:10.034+0100 [redisson-5-7] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 33786536 ns ago, task: org.redisson.QueueTransferTask$3@3eb91594)
2017-10-31T16:38:10.034+0100 [redisson-5-7] INFO org.redisson.QueueTransferTask - New timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 965997362 ns later, task: org.redisson.QueueTransferTask$3@2c65b5ba)
2017-10-31T16:38:10.034+0100 [redisson-netty-4-6] INFO org.redisson.QueueTransferTask - Cancel timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 965669575 ns later, task: org.redisson.QueueTransferTask$3@2c65b5ba)
2017-10-31T16:38:10.035+0100 [redisson-netty-4-6] INFO org.redisson.QueueTransferTask - New timeout: HashedWheelTimer$HashedWheelTimeout(deadline: 26509964992835 ns later, task: org.redisson.QueueTransferTask$3@4bb12315)
```

I added newlines to separate "seconds". After the last line of this log, there is no more logs. So the bug is reproduced.

Here how I modified the `scheduleTask` with some logs :

```
    private void scheduleTask(final Long startTime) {
        if (startTime == null) {
            return;
        }
        
        Timeout oldTimeout = timeoutReference.get();
        if (oldTimeout != null) {
            log.info("Cancel timeout: {}", oldTimeout);
            oldTimeout.cancel();
            timeoutReference.compareAndSet(oldTimeout, null);
        }
        
        long delay = startTime - System.currentTimeMillis();
        if (delay > 10) {
            Timeout timeout = connectionManager.newTimeout(new TimerTask() {
                @Override
                public void run(Timeout timeout) throws Exception {
                    log.info("Timeout expired: {}", timeout);
                    pushTask();
                }
            }, delay, TimeUnit.MILLISECONDS);
            log.info("New timeout: {}", timeout);
            timeoutReference.set(timeout);
        } else {
            pushTask();
        }
    }
```

What we can see is that the 2 first blocks in the log are working well.
The order of calls in the last block is different, and we can see that the next timeout for the every seconds task is **canceled** when the new timeout for every days at midnight task.

In the two first blocks, this "cancelation" occurs before the new timeout for every seconds task is scheduled.
So I was false when I was saying that the problem was from the `HashedWheelTimer`.

The problem is that the order in which things occurs is responsible of the bug.

My question is : is this normal that the order of things is responsible of the bug, or not.
In other words, is the algorithm good in this method with the AtomicReference of the old timeout, or not.

Remember that I schedule 2 cron tasks on the same executor service.

Thanks.

**Comment by jackygurui** [comment]
Thanks for your detailed research. Shall look at this issue in more detail.

**Comment by mrniko** [comment]
Fixed!

@rlamarche @jackygurui Thanks for help with research of this issue!

**Comment by rlamarche** [comment]
Thank you @mrniko and @jackygurui for your support !
I'm not able to reproduce the bug with the fix.

:+1:

**Comment by jackygurui** [comment]
Hats off to @rlamarche for the extraordinarily detailed research to nail the bug.
